autoparallel/api.py

```diff
@@ -504,30 +547,82 @@ def forward(self, *args):
                 return out
 
         self.parallel_model = AutoParallelModule()
+        self._register_params_and_init_weights(sharded_param_dict, sharded_buffer_dict)
+        return self.parallel_model
 
-        # We construct an unflattened structure on parallel_mod,
-        # e.g. _assign_attr(v, parallel_model, k="layers.0.weight") will literally
-        # create empty nn.Modules recursively and then stash 'v' so it shows up in the right spot
-        for k, v in sharded_param_dict.items():
-            _assign_attr(v, self.parallel_model, k, attr_kind=_AttrKind.PARAMETER)
-
-        for k, v in sharded_buffer_dict.items():
-            _assign_attr(v, self.parallel_model, k, attr_kind=_AttrKind.BUFFER)
-
-        # Right now we require a convention that the user model provides an init_weights method,
-        # although we could snoop for other methods too.
-        hook_params_setters(self.init_weights_model, self.parallel_model)
-        if hasattr(self.model, "init_weights"):
+    def apply_placement_pp(self, sharding_placement=None) -> torch.nn.Module:
+        sharded_param_dict, sharded_buffer_dict = self._apply_placement_common(
+            sharding_placement
+        )
+        from autoparallel._passes.graph_partition import (
+            partition_joint_with_descriptors,
+        )
 
-            def init_weights(_self, *args, **kwargs):
-                # this is now a deep-fake-copy of orig mod, so we don't have to use reparametrize
-                return self.init_weights_model.init_weights(*args, **kwargs)
+        (
+            fw_module,
+            bw_module,
+            num_user_outputs,
+            num_mutate_inputs,
+            num_fw_outs_saved_for_bw,
+            num_symints_saved_for_bw,
+            _indices_of_inps_to_detach,
+            adjusted_flat_args,
+        ) = partition_joint_with_descriptors(self.joint_with_descriptors)
+
+        class AutoParallelPPStage(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, *args):
```

Comment:
one question (maybe i'm missing something obvious) - how do we plan on using this custom `autograd.Function` in the pipeline runtime? I would have imagined that instead of custom autograd.Functions, we just have a bunch of `fx.GraphModules` floating around that the pipeline runtime invokes one after the other